### PR TITLE
Upgrade yeoman-generator to version 0.19.1

### PR DESCRIPTION
**Pull request by alfonso-presa**
I've tried to upgrade the generator to the 0.19.1 API as it was using fs directly.

Let me know if there's something wrong.

BTW, I tried to use sinom to spy on the generator.log.error method for testing asserts but it seems to be already wrapped.

### CODE REVIEW

**Comment by tomwayson** [comment]
@alfonso-presa we should probably also replace calls to `this.template()` w/ `this.fs.copyTmpl()` right?

**Comment by tomwayson** [comment]
@alfonso-presa 

I made a PR to your UpgradeGeneratorVersion branch which includes a couple of extra changes that I *think* help bring this more in line w/ the way things are supposed to work in yeoman 0.19, but I'll let @SBoudrias weigh in on whether or not I've got it right:

https://github.com/alfonso-presa/generator-karma/pull/1

**Comment by alfonso-presa** [comment]
Included @tomwayson changes and @SBoudrias comments to his PR. Thank you guys.

File: app/index.js [hunk 1]
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
...
+        this.options.plugins.push('coffee-script');
       }
 
-      try {
-        data = JSON.parse(content);
-      } catch (err) {
-        this.log.error('Could not parse package.json.', err);
-        done();
+      var data = this.fs.readJSON(this.destinationPath('package.json'));
+      if(!data) {
```
Comment (kevva) [inline app/index.js]: `if (!data)`

File: app/index.js [hunk 2]
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
...
+  writing: {
+    makeConfig: function () {
+      this.sourceRoot(path.join(__dirname, this.options['template-path']));
+
+      this.fs.copyTpl(
+        this.templatePath(this.options['config-file']),
+        this.destinationPath(path.join(this.options['config-path'], this.options['config-file'])),
+        {
+          pkg: this.pkg,
+          basePath: this.options["base-path"],
```
Comment (SBoudrias) [inline app/index.js]: Use single quote.

File: app/index.js [hunk 3]
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
...
+        this.log.error('Could not open package.json for reading.');
         return;
       }
 
       data.devDependencies = data.devDependencies || {};
-      this.options.plugins.forEach(function (plugin) {
-        data.devDependencies[plugin] = '*';
-      })
-      data.devDependencies = sortedObject(data.devDependencies);
+        this.options.plugins.forEach(function (plugin) {
```
Comment (SBoudrias) [inline app/index.js]: This indentation looks wrong, seems to be one off.

File: app/index.js [hunk 4]
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
...
-        return;
+        this.log.writeln('Test script already present in package.json. Skipping rewriting.');
       }
-
-      data.scripts = data.scripts || {};
-      data.scripts.test = 'grunt test';
-
-      fs.writeFile(packageJson, JSON.stringify(data, null, 2), done);
-    }.bind(this));
+      else{
```
Comment (SBoudrias) [inline app/index.js]: The correct structure of if/else block according to our styleguide is:

```js
if () {

} else {

}
```

File: app/index.js [hunk 5]
```diff
@@ -166,144 +165,120 @@ module.exports = yeoman.generators.Base.extend({
...
+      }
+    }
   },
 
-  installDeps: function () {
-    if (!this.options['skip-install']) {
-        this.npmInstall();
-    }
+  install: function () {
+    this.npmInstall(undefined,undefined,this.async());
```
Comment (SBoudrias) [inline app/index.js]: Don't pass any argument here.

Also note that passing `this.async()` callback here will deadlock the node process and just close the generator. - So this is actually breaking the generator :P

File: test/test-creation.js [hunk 6]
```diff
@@ -50,24 +50,18 @@ describe('Karma generator creation test', function () {
...
-
-      afterEach(function () {
-        process.stderr.write = stderrWriteBackup;
-      });
-
       it('creates a travis file but warns about missing package.json', function(done) {
-        gen.withOptions({travis: true}).on('end', function () {
+        var logMessage = '';
+        gen.withOptions({travis: true})
+        .on('ready', function() {
```
Comment (SBoudrias) [inline test/test-creation.js]: Indentation of this block seems off. It would need one more.

File: test/test-creation.js [hunk 7]
```diff
@@ -75,31 +69,34 @@ describe('Karma generator creation test', function () {
...
-    helpers.testDirectory(join('test', 'temp'), function () {
-      var gen = helpers.run(join(__dirname, '../app'));
-      require('fs').writeFileSync('package.json', '{}');
-      gen.withOptions({travis: true}).on('end', function () {
-        assert.file(['.travis.yml']);
-        assert.fileContent('package.json', /grunt test/);
-        done();
-      });
+    helpers
+    .run(join(__dirname, '../app'))
```
Comment (SBoudrias) [inline test/test-creation.js]: Indentation of these blocks seems off. It would need one more.

**Comment by SBoudrias** [comment]
Added some comments about the code style and a bug I found.

**Comment by alfonso-presa** [comment]
@SBoudrias, thanks for the feedback. They should all be corrected in the last commit. 

About the async()... I was expecting this to be necessary for the generator not to trigger the "end" event until the instalations have ended. In the past I've seen the "npm install" running at the same time when using one generator inside another (i.e. generator-angular) causing concurrency issues....is this solved in generator-yeoman 0.19?

**Comment by SBoudrias** [comment]
It's been solved for a couple versions now. But the deadlocking issue is only in yeoman-generator since 0.18

**Comment by alfonso-presa** [comment]
Ok, thanks! Great to hear it's solved :-). I removed the async call in the last commit, so I guess everything is ok now.

File: app/index.js [hunk 8]
```diff
@@ -2,13 +2,12 @@
...
-var fs = require('fs');
-var sortedObject = require("sorted-object");
+var sortedObject = require('sorted-object');
 
 var _ = require('underscore');
 
 module.exports = yeoman.generators.Base.extend({
-  init: function () {
+  initializing: function () {
     var notEmpty = function (str) {
```
Comment (SBoudrias) [inline app/index.js]: Can we rename this function to something explicit? `notEmpty` sounds like a boolean check, not an array filtering method...

Maybe `stringArray` or `arrayFromString`

File: app/index.js [hunk 9]
```diff
@@ -166,144 +165,119 @@ module.exports = yeoman.generators.Base.extend({
...
-        done();
+      this.fs.copy(
+        this.templatePath('travis.yml'),
+        this.destinationPath('.travis.yml')
+      );
+
+      var data = this.fs.readJSON(this.destinationPath('package.json'));
+
+      if (!data) {
+        this.log.error('Could not open package.json for reading.', {});
```
Comment (SBoudrias) [inline app/index.js]: no need for the empty object `{}`

File: test/test-creation.js [hunk 10]
```diff
@@ -50,56 +50,53 @@ describe('Karma generator creation test', function () {
...
 
   it('creates a travis file and updates package.json', function(done) {
-    helpers.testDirectory(join('test', 'temp'), function () {
-      var gen = helpers.run(join(__dirname, '../app'));
-      require('fs').writeFileSync('package.json', '{}');
-      gen.withOptions({travis: true}).on('end', function () {
+    helpers
+      .run(join(__dirname, '../app'))
+      .withOptions({travis: true, force: true})
+      .inDir(join(__dirname, 'temp'), function(dir){
```
Comment (SBoudrias) [inline test/test-creation.js]: `function () {`

Comment (SBoudrias) [inline test/test-creation.js]: let's use `inTmpDir()`

**Comment by SBoudrias** [comment]
Some new comments, went through the whole thing this time so it should be the last iteration of review comments :)

File: app/index.js (hunk 8 above)
Comment (addyosmani) [inline app/index.js]: :+1: arrayFromString

**Comment by addyosmani** [comment]
Minus the code style comments, LGTM2.

**Comment by alfonso-presa** [comment]
Sorry for the delay... here you have the changes.

By the way.. I've seen that when using this inside other generators based in older versions of yeoman-generator (for example generator-angular) it now displays conflict when generator-karma is editing package.json (even if the folder was initially empty).... Is there any way to solve this different than upgrading that generator to 0.19.x?

**Comment by alfonso-presa** [comment]
Looks like using inTmpDir breaks the build in travis due to a failing test. My guess is that in travis it's using the same tmp folder and it's not clearing it before running the tests. This is the failing build:
https://travis-ci.org/yeoman/generator-karma/builds/59006966

Changing back to inDir solved the issue...

**Comment by alfonso-presa** [comment]
Ok... it was my fault. For some reason in travis (probably because it's slower than my machine) it was ending the inTmpDir before it appends the onready event that adds the spy to the log error in the test. I rewrote the test to ensure the onready is attached.

I have already squashed the commits in case you wan't to merge.

**Comment by SBoudrias** [comment]
Thanks a lot!!

**Comment by alfonso-presa** [comment]
Thank you!

**Pull request merged**